features/org.eclipse.equinox.executable.feature/library/eclipse.c

```diff
@@ -650,32 +653,35 @@ static int _run(int argc, _TCHAR* argv[], _TCHAR* vmArgs[])
 		case 0: /* normal exit */
 			break;
 		case RESTART_LAST_EC:
-			/* copy for relaunch, +1 to ensure NULL terminated */
-			relaunchCommand = malloc((initialArgc + 1) * sizeof(_TCHAR*));
-			memcpy(relaunchCommand, initialArgv, (initialArgc + 1) * sizeof(_TCHAR*));
-			relaunchCommand[initialArgc] = 0;
-			relaunchCommand[0] = program;
-			break;
-
+			if (!noRestart) {
+				/* copy for relaunch, +1 to ensure NULL terminated */
+				relaunchCommand = malloc((initialArgc + 1) * sizeof(_TCHAR*));
+				memcpy(relaunchCommand, initialArgv, (initialArgc + 1) * sizeof(_TCHAR*));
+				relaunchCommand[initialArgc] = 0;
+				relaunchCommand[0] = program;
+				break;
```

Comment:
the break of case statment should still be outside of if condition, right? otherwise any data put in exitData buffer by the application will be considered error message in `default` case..

Reply:
The essence is:

```java
		case RESTART_LAST_EC:
			if (!noRestart) {
                                ...
				break;
			} // else: fall-through to 'default' case.
		case RESTART_NEW_EC:
			if (!noRestart) {
                                ...
				break;
			} // else: fall-through to 'default' case.
		default: {
			...
```

If the exit code is a restart exit code, then if we allow restart (we do not disallow restarting), we do something special (whatever was there already before) and then break. But, if restart is disabled, we instead fall through to the `default` case, and handle it like any other non-zero exit code.

So, that part still looks good to me.

Reply:
and this is infact the reason of tests failure..

test app set exit data here if case from within application user select restart.

https://github.com/eclipse-equinox/equinox/blob/99a85789ed29f87d4ed04d15bb640de325f4750f/bundles/org.eclipse.equinox.launcher.tests/src/org/eclipse/equinox/launcher/TestLauncherApp.java#L89

and in launcher, as `-norestart` is mentioned and `default` is executed, this exitData is shown as error..

https://github.com/eclipse-equinox/equinox/blob/1551906df8e01c74351712db0d602495458fa7fd/features/org.eclipse.equinox.executable.feature/library/eclipse.c#L735

Reply:
I was trying to only skip the `RESTART_*` cases, but I think you're saying also `exitData` may contain something related to restart behavior? I'm really not sure what that variable is about, at all.

`eclipse.c` has more than 2000 lines. I have to say there is a lot of stuff going on there, and I don't know what half of it means. It is complicated.

Can the `default` case really be skipped fully? It seems to handle reporting various errors. For instance, the `errorMsg == NULL` seems relevant, as it reports problems launching Java? Are you sure we can safely skip all of that? I don't want it to print anything in case of a non-zero exit code, but it should probably still report issues for failing to launch Java to begin with.

Should we do `if (exitData != NULL) free(exitData);` before falling through? And what is `getSharedData` about? That can also set `exitData` again?

Reply:
IMO moving `break` out of `if` conditions will fix the issue..

Reply:
> IMO moving `break` out of `if` conditions will fix the issue..

Yes, I got that impression as well. But, while that may fix the test, will that not prevent too many other errors from being reported as well (see my previous comment).

Reply:
it would become as first case where application is expected to just exit..

https://github.com/eclipse-equinox/equinox/blob/2ef1133fd652c1aef50e41a0b3369ad6ed11eb81/features/org.eclipse.equinox.executable.feature/library/eclipse.c#L653-L654

Reply:
Yes, I get that. But is that desired? I don't think users would expect that if they disable restart behavior, that they suddenly also get errors for failing to start Java reported in a different way. It would fit my use case just fine, but I'm not sure it is really logical for others?

Reply:
>  if they disable restart behavior, that they suddenly also get errors for failing to start Java reported in a different way.

this will not happen if you break from all cases.. in which code flow you think this will happen?

Reply:
If `javaResults->launchResult` is `23` or `24` and `javaResults->runResult` is zero, we'd get into the restart cases and fall through to the `default` (assuming the keep the `break`s where they are now). If `launchMode != LAUNCH_EXE` and `exitData == 0`, we'd go into `errorMsg == null` and since `javaResults->runResult` is not taken as branch in the `if`, and assuming `javaResults->errorMessage != NULL`, we set `errorMsg`, and just below we'd print that. By moving the `break` statements, and thus not falling through to the `default` case, this is not reported anymore. I think it is still valuable to report issues in launching Java, as otherwise a normal exit code can't be distinguished anymore from a failure to launch Java.

Reply:
@umairsair What do you think of these cases?

Reply:
> assuming javaResults->errorMessage != NULL, we set errorMsg, and just below we'd print that. 

the code comment on it says that we have a problem launching java, which means that in this flow user app will not have been run and not returned 23 or 24.. we need to completely follow the same code flow as before, the only change we are doing in code flow is to not restart java/jvm if --norestart is set and it'll be achieved by simply conditioning the code that is doing the restart and let the remaining code flow same, and it would be achieved if you simply put case 23 and 24 code under condition except break statement..

BTW did you try running tests after moving break out of condition? does it solve the test failures?

Reply:
* I've moved the `break` statements as suggested by @umairsair. Let's see if that works. Could someone allow another build?
* I've rebased on `master` and squashed all changes to a single commit, as requested.

Reply:
Build failed:

```
Failed to execute goal org.eclipse.tycho.extras:tycho-p2-extras-plugin:4.0.8:compare-version-with-baselines (compare-attached-artifacts-with-release) on project org.eclipse.equinox.app: Only qualifier changed for (org.eclipse.equinox.app/1.7.100.v20240608-1512). Expected to have bigger x.y.z than what is available in baseline (1.7.100.v20240321-1445) -> [Help 1]
```

I tried changing the version in the manifest of `org.eclipse.equinox.app` from `1.7.100.qualifier` to `1.7.101.qualifier`, but that gives a warning: `The service version is increased unnecessarily since either the major or minor or service version is already increased`.

What should I do to get the build working?

Reply:
update version to `1.7.200.qualifier` at following location

https://github.com/eclipse-equinox/equinox/blob/5851137b9bbfad795a9f0c00c5646f9aa80bde78/bundles/org.eclipse.equinox.app/META-INF/MANIFEST.MF#L5

Reply:
I changed the version.

Reply:
I also changed the feature, as it then complained about that. It seems the latest build was 'aborted'. Could someone approve or restart the build/test?

Reply:
I've retriggered the build

Reply:
I had to recursively adapt 4 more features. I hope that is all the versions that need updating. Needs another build approval.

Reply:
If I understand it correctly, the new tests pass. There is one test that fails: `testCoordinatedConfigurationOnBeforeRegisteredManagedService (org.osgi.test.cases.cm.junit.CMCoordinationTestCase) failed. Not sure what that is, or where it comes from, and whether that is even related to the changes I made? Also, 2/23 checks were aborted. I also have no idea why that is.

Reply:
The `CMCoordinationTestCase` failed test also fails in #640 that @HannesWell [refers to](https://github.com/eclipse-equinox/equinox/pull/595#issuecomment-2164057367) in a different thread of this PR. I don't think that failure is related to my changes.

I still don't know what caused the 2 other checks to be aborted. Their output doesn't help me much. Although, I see they were 'aborted'; they didn't 'fail'. I also in Eclipse ESCET recently got timeout aborts due to the Jenkins cluster being slower than usual, and I had to increase the timeout there. Could that be the reason here as well?

Reply:
I have not yet read to this entire thread, but my first impression when looking into the change itself is that in case a no-restart is present the code should fall through to the default branch and tread the exit code as error (which it in fact is, isn't it)?

By default the executable pops-up a native error dialog if the app exited with an error code. 
Could the test-failure with a fall-through to default be because of that? Because at the server such pop-up is closed by nobody.
Maybe the test would be more correct if you use from the [Eclipse runtime options](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fruntime-options.html)
```
--launcher.suppressErrors (Executable)
    If specified the executable will not display any error or message dialogs.  This is useful if the executable is being used in an unattended situation.
```
Then the error code is just returned by the command and could be checked in the invoking java test code?